File: rules-interpreter/src/test/java/RouterTest.java
import io.github.iamrenny.ruleflow.Workflow;
import io.github.iamrenny.ruleflow.vo.Action;
import io.github.iamrenny.ruleflow.vo.WorkflowResult;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;
import java.util.Set;

class RouterTest {

    @Test
    public void givenRoutingWorkflowWhenRoutingToDefaultCaseMustReturnValuesSuccessfully() {
        String workflow = """
          workflow 'providers'
          ruleset 'routing' 
               'provider1_master1' user_id = '101000042449' return route with route({'provider':'provider1', 'merchant_id': 'master1'})
               'provider1_master2' user_id = '151000033978'  return route with route({'provider': 'provider1', 'merchant_id': 'master1'})
               'provider2_master1' user_id = '151000033979' return route with route({'provider': 'provider2', 'merchant_id': 'master3'})
               'provider3' user_id = '151000033977' return route with route({'provider': 'provider3', 'merchant_id': 'merchant4'})
               'provider4' user_id = '15100003391381823' return route with route({'provider': 'provider4', 'tag': 'merchant5'})
               'provider5' user_id = '151000033976' return route with route({'provider': 'internal'})
           default route with route({'provider': 'internal'})
           end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult(
            "providers",
            "default",
            "default",
            "route",
            Map.of("route", Map.of("provider", "internal"))
        );

        WorkflowResult result = ruleEngine.evaluate(Map.of("user_id", "4"));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenRoutingWorkflowWhenRoutingToSpecificRuleCaseMustReturnValue() {
        String workflow = """
          workflow 'router'
               ruleset 'providers' 
               'provider1' country in 'ar' return route with route({'provider':'provider1', 'merchant_id': 'master'})
               'provider2' country in 'us' return route with route({'provider': 'provider2', 'merchant_id': 'master2'})
               'provider3' user.type = 'premium' OR user_is_blacklisted = false return route with route({'provider': 'provider3', 'merchant_id': 'master3'})
               'provider4' order.vertical in 'restaurants', 'travel' return route with route({'provider': 'provider4', 'merchant_id': 'master4'})
           default do_not_route
           end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult(
            "router",
            "providers",
            "provider2",
            "route",
            Map.of("route", Map.of("provider", "provider2", "merchant_id", "master2"))
        );

        WorkflowResult result = ruleEngine.evaluate(Map.of("country", "us"));

        Assertions.assertEquals(expectedResult, result);
    }
}

File: rules-interpreter/src/test/kotlin/RouterTest.kt
import com.github.iamrenny.rulesflow.Workflow
import com.github.iamrenny.rulesflow.vo.Action
import com.github.iamrenny.rulesflow.vo.WorkflowResult
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class RouterTest {

    @Test
    fun `given a routing workflow, when routing to default case must return values successfully`() {
        val workflow = """
          workflow 'providers'
          ruleset 'routing' 
               'Cybersourse1-rappi_master' user_id = '101000042449' return route with route({'provider':'cybersource', 'merchant_id': 'rappi_master2'})
               'Cybersourse2-rappi_master2' user_id = '151000033978'  return route with route({'provider': 'cybersource', 'merchant_id': 'rappi_master2'})
               'Cybersourse3-rappi_dm_bwl' user_id = '151000033979' return route with route({'provider': 'cybersource', 'merchant_id': 'rappi_dm_bwl'})
               'Cybersourse4-rappi_dm_rest_5a15_tkt' user_id = '151000033977' return route with route({'provider': 'cybersource', 'merchant_id': 'rappi_dm_rest_5a15_tkt'})
               'Riskified-rappi' user_id = '15100003391381823' return route with route({'provider': 'riskified', 'tag': 'rappi'})
               'Fraud_rules_engine' user_id = '151000033976' return route with route({'provider': 'fraud_rules_engine'})
           default route with route({'provider': 'fraud_rules_engine'})
           end
        """
        val ruleEngine = Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult(
                "providers",
                "default",
                "default",
                "route",
                warnings = setOf(),
                actions = setOf("route"),
                actionsWithParams = mapOf("route" to mapOf("provider" to "fraud_rules_engine")),
                actionsList = listOf(Action("route", mapOf("provider" to "fraud_rules_engine")))
            ),
            ruleEngine.evaluate(
                mapOf(
                    "user_id" to "4"
                )
            )
        )
    }

    @Test
    fun `given a routing workflow, when routing to a specific rule case must return value`() {
        val workflow = """
          workflow 'router'
                   ruleset 'providers' 
                   'Cybersourse1-rappi_master' country in 'ar', 'cl', 'co', 'ec', 'pe' return route with route({'provider':'cybersource', 'merchant_id': 'rappi_master'})
                   'Cybersourse3-rappi_master2' country in 'br', 'cr', 'mx', 'uy' return route with route({'provider': 'cybersource', 'merchant_id': 'rappi_master2'})
                   'Cybersourse4-rappi_dm_bwl' user_is_whitelisted = true OR user_is_blacklisted = true return route with route({'provider': 'cybersource', 'merchant_id': 'rappi_dm_bwl'})
                   'Cybersourse4-rappi_dm_rest_5a15_tkt' order.vertical in 'restaurants', 'travel' return route with route({'provider': 'cybersource', 'merchant_id': 'rappi_dm_rest_5a15_tkt'})
           default do_not_route
           end
        """
        val ruleEngine = Workflow(workflow)
        Assertions.assertEquals(
            WorkflowResult(
                "router",
                "providers",
                "Cybersourse3-rappi_master2",
                "route",
                warnings = setOf(),
                actions = setOf("route"),
                actionsWithParams = mapOf("route" to mapOf("provider" to "cybersource", "merchant_id" to "rappi_master2")),
                actionsList = listOf(Action("route", mapOf("provider" to "cybersource", "merchant_id" to "rappi_master2")))
            ),
            ruleEngine.evaluate(
                mapOf(
                    "country" to "br"
                )
            )
        )
    }
}